src/main/java/bio/terra/service/tabulardata/google/BigQueryPdao.java

```diff
@@ -870,6 +1026,50 @@ private String sourceSelectSql(String snapshotId, Column targetColumn, SnapshotM
         }
     }
 
+    // for each table in a dataset (source), collect row id matches ON the row id
+    public RowIdMatch matchRowIds(Snapshot snapshot, SnapshotSource source, String tableName, List<String> rowIds) {
+        // One source: grab it and navigate to the relevant parts
+        BigQueryProject bigQueryProject = bigQueryProjectForSnapshot(snapshot);
+
+        Optional<SnapshotMapTable> optTable = source.getSnapshotMapTables()
+            .stream()
+            .filter(table -> table.getFromTable().getName().equals(tableName))
+            .findFirst();
+        // create a column to point to the row id column in the source table to check that passed row ids exist in it
+        Column rowIdColumn = new Column()
+            .table(optTable.get().getFromTable())
+            .name(PDAO_ROW_ID_COLUMN);
+
+
+        ST sqlTemplate = new ST(mapValuesToRowsTemplate);
+        sqlTemplate.add("project", bigQueryProject.getProjectId());
```

Comment:
Since these `sqlTemplate.add` sections pop up a bunch, I wonder if it would be valuable to refactor them (and potentially clarify which items are needed when)

Reply:
Personally, I think it would be clearer to keep the template text adjacent to the code that fills it in. Otherwise, you can't see the correlation of parameters.

Reply:
That's true. I'll leave this one out.